README.md

```diff
@@ -1 +1,235 @@
+
+- // remove environment constants that are defined in the codebase or at compile time
+- import { environment } from './environments/environment';
+
+  @Injectable()
+  export class DataService {
+    configUrl = ;
+
+    constructor(
+        private http: HttpClient
+        private
+        ) { }
+
+    getData() {
+-      return this.http.get(`${environment.API}/assets/data.json`);
++      // use globally scoped environment variables that are unique to the deployment
++      return this.http.get(`${window.env.API}/assets/data.json`);
+    }
+  }
+```
+
+#### Static assets must be hosted at locations that are unique and _independent of the web application environment(s)_.
+
+Static assets that do not contain anything environment-specific can be built once, published to once to a unique location, and then used in multiple environments of the web application.
+
+These static assets share the same qualities as hosted javascript libraries on content delivery networks (CDN) ([Google Hosted Libraries](https://developers.google.com/speed/libraries/), [cdnjs](https://cdnjs.com/), [jsDelivr](https://www.jsdelivr.com/), [UNPKG](https://unpkg.com/)):
+
+<code>https:/<span></span>/<span style="font-weight: bold">ajax.googleapis.c<span></span>om</span>/ajax/libs/jquery/<span style="font-weight: bold">3.3.1</span>/jquery.min.js</code>
+
+The location of the jquery library above, referenced by innumerable web applications, is both independent of the applications and uniquely versioned.
+
+<code>https:/<span></span>/<span style="font-weight: bold">assets.myapp.c<span></span>om</span>/apps/<span style="font-weight: bold">1.0.2</span>/main.js</code>
+
+Similarly, the location of the example the web application javascript file is unique and hosted at a location that is dedicated to static assets. The static asset web server is a repository for versions of the web application.
+
+#### Configure the static assets for long-term caching
+
+Static assets that do not contain anything environment-specific and  are hosted at a unique and permanent location may be configured to be [cached by the browser (almost) indefinitely](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#Caching_static_assets):
+
+ `cache-control: public, max-age=31536000, immutable`
+
+### `index.html` is deployable configuration
+
+ The HTML document of a single-page application (often `index.html`) is ___not___ static. It varies from environment to environment, and deploy to deploy. The HTML document is a composition of the environment-specific configuration and _immutable_ static assets that define the web application.
+
+#### `index.html` contains fully-qualified references to the static assets
+
+```diff
+- <!-- not unique, not independent -->
+- <script src="main.js" type="text/javascript"></script>
+
++ <!-- permanent, reuseable -->
++ <script src="https://assets.myapp.com/apps/bae1407/main.js" type="text/javascript"></script>
```

Comment:
What is bae1407??  Might be better to carry through the same example asset name/version that you refer to earlier in the doc.